optuna/study.py

```diff
@@ -269,19 +268,26 @@ def optimize(
                 memory is safely managed in your objective function.
         """
 
-        if not self._optimize_lock.acquire(False):
-            raise RuntimeError("Nested invocation of `Study.optimize` method isn't allowed.")
         if not isinstance(catch, tuple):
             raise TypeError("The catch argument is of type \'{}\' but must be a tuple.".format(
                 type(catch).__name__))
 
+        if not self._optimize_lock.acquire(False):
+            raise RuntimeError("Nested invocation of `Study.optimize` method isn't allowed.")
+
         try:
             if n_jobs == 1:
                 self._optimize_sequential(func, n_trials, timeout, catch, callbacks,
-                                          gc_after_trial)
+                                          gc_after_trial, None)
             else:
-                self._optimize_parallel(func, n_trials, timeout, n_jobs, catch, callbacks,
-                                        gc_after_trial)
+                time_start = datetime.datetime.now()
+                with Parallel(n_jobs=n_jobs, prefer="threads") as parallel:
+                    _iter = range(n_trials) if n_trials is not None else iter(int, 0)
```

Comment:
If `n_trials` is `None`, the study doesn't run any trials because `iter(int, 0)` returns an empty iterator.

How about fixing the code to create `_iter` as follows:
```python
if n_trials is not None:
    _iter = range(n_trials)
elif timeout is not None:
    is_timeout = lambda: (datetime.datetime.now() - time_start).total_seconds() > timeout
    _iter = iter(is_timeout, True)
else:
    # The following expression makes an iterator that never ends.
    _iter = iter(int, 1)
```

Reply:
Nice catch. I did not test that part of the code.
Thanks for the suggested fix. I'll add it right away.